controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java

```diff
@@ -0,0 +1,115 @@
+ * Auth manager class for Pravega controller. This manages the handlers for grpc and REST together.
+ * In case of grpc, the routing of the authenticate function to specific registered interceptor is taken care by grpc
+ * interceptor mechanism.
+ * In case of REST calls, this class routes the call to specific AuthHandler.
+ */
+@Slf4j
+public class PravegaAuthManager {
+    private final GRPCServerConfig serverConfig;
+    @GuardedBy("this")
+    private final Map<String, AuthHandler> handlerMap;
+
+    public PravegaAuthManager(GRPCServerConfig serverConfig) {
+        this.serverConfig = serverConfig;
+        this.handlerMap = new HashMap<>();
+    }
+
+    private AuthHandler getHandler(String handlerName) throws AuthenticationException {
+        AuthHandler retVal;
+        synchronized (this) {
+                retVal = handlerMap.get(handlerName);
+            }
+            if (retVal == null) {
+            throw new AuthenticationException("Handler does not exist for method " + handlerName);
+        }
+        return retVal;
+    }
+
+    /**
+     * API to authenticate and authroize access to a given resource.
+     * @param resource The resource identifier for which the access needs to be controlled.
+     * @param headers  Custom headers used for authentication.
+     * @param level    Expected level of access.
+     * @return         Returns true if the entity represented by the custom auth headers had given level of access to the resource.
+     * @throws AuthenticationException Exception faced during authentication/authorization.
+     */
+    public boolean authenticate(String resource, MultivaluedMap<String, String> headers, AuthHandler.Permissions level) throws AuthenticationException {
+        Map<String, String> paramMap = headers.entrySet().stream().collect(Collectors.toMap(k -> k.getKey(), k -> k.getValue().get(0)));
+        return authenticate(resource, paramMap, level);
+    }
+
+    /**
+     * API to authenticate and authroize access to a given resource.
+     * @param resource The resource identifier for which the access needs to be controlled.
+     * @param paramMap  Custom headers used for authentication.
+     * @param level    Expected level of access.
+     * @return         Returns true if the entity represented by the custom auth headers had given level of access to the resource.
+     * @throws AuthenticationException Exception faced during authentication/authorization.
+     */
+    public boolean authenticate(String resource, Map<String, String> paramMap, AuthHandler.Permissions level) throws AuthenticationException {
+        boolean retVal = false;
+        try {
+            String method = paramMap.get("method");
+            AuthHandler handler = getHandler(method);
+            retVal = handler.authenticate(paramMap) &&
+                    handler.authorize(resource, paramMap).ordinal() >= level.ordinal();
+        } catch (RuntimeException e) {
+            throw new AuthenticationException(e);
```

Comment:
I think it's important to dissociate between `authenticate` failing which should result in a http 401, and `authorize` failure, which should result in an http 403.   Right now, failure of any causes 401.